# Homework 2
- Họ và tên: Nguyễn Phúc Khang
- MSSV: 22127180

## Import các thư viện cần thiết

In [1]:
import datetime as dt # Để làm việc với dữ liệu datetime
import math

## Đưa ra câu hỏi cần trả lời

Tình trạng hiểu về Python các bạn trong lớp?

## Thu thập dữ liệu

Cho làm quiz về Python trên moodle.

Kết quả có thể được download về dưới dạng file .csv.

In [2]:
## Khám phá dữ liệu

# Đọc file và encoding để không lỗi định dạng
file = open('Data/PythonQuiz_2024_CLC.csv', 'r', encoding='utf-8-sig')

# Xử lý định dạng
# string loại bỏ khoảng trắng 2 đầu, \n, \t,...
# split tách từ theo 'char'
first_line = file.readline().strip().split(',')

# Khởi tạo Dictionary để lưu dữ liệu key:value - feature:[data]
cols = {}
for val in first_line:
    cols[val] = []

for line in file:
    line_vals = line.strip().split(',')
    if line_vals[0] == 'Never submitted':
        continue

    for i in range(len(line_vals)):
        cols[first_line[i]].append(line_vals[i])

cols

{'State': ['Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished',
  'Finished'],
 'Started on': ['16/10/2024 17:14',
  '16/10/2024 18:59',
  '16/10/2024 19:06',
  '16/10/2024 21:27',
  '16/10/2024 23:15',
  '17/10/2024 0:04',
  '17/10/2024 0:19'

### Dữ liệu có bao nhiêu dòng và bao nhiêu cột?

In [3]:
cols.keys()

dict_keys(['State', 'Started on', 'Completed', 'Time taken', 'Grade/10.00', 'Q. 1 /0.53', 'Q. 2 /0.53', 'Q. 3 /0.53', 'Q. 4 /0.53', 'Q. 5 /0.53', 'Q. 6 /0.53', 'Q. 7 /0.53', 'Q. 8 /0.53', 'Q. 9 /0.53', 'Q. 10 /0.53', 'Q. 11 /0.53', 'Q. 12 /0.53', 'Q. 13 /0.53', 'Q. 14 /0.53', 'Q. 15 /0.53', 'Q. 16 /0.53', 'Q. 17 /0.53', 'Q. 18 /0.53', 'Q. 19 /0.53'])

In [4]:
# Số dòng
len(cols['State'])

59

In [5]:
# Số cột
len(cols)

24

### Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng ứng với kết quả làm quiz của một sinh viên.

Xem qua bảng dữ liệu của file csv thì có vẻ không thấy có dòng nào "lạc loài" (thật ra dữ liệu ban đầu là có dòng cuối là dòng tính điểm trung bình của các cột, nhưng mình đã bỏ đi rồi).

### Dữ liệu có các dòng bị lặp không?

Xem qua bảng dữ liệu của file csv thì có vẻ không thấy có vấn đề một sinh viên xuất hiện nhiều hơn một lần.

### Mỗi cột có ý nghĩa gì?

- Cột "State": cho biết trạng thái làm bài, chẳng hạn "Finished" là đã nộp bài (đã ấn nút "submit all and finish")
- 3 cột "Started on", "Completed", "Time taken": lần lượt cho biết thời điểm bắt đầu làm, thời điểm làm xong, và thời gian từ khi bắt đầu làm cho tới khi làm xong
- Cột "Grade/16.00": cho biết điểm tổng (tối đa là 16, vì có 16 câu và mỗi câu 1 điểm)
- 16 cột còn lại "Q. 1 /1.00", "Q. 2 /1.00", ...: cho biết điểm của mỗi câu quiz (mỗi câu làm đúng hoàn toàn thì sẽ được 1 điểm)

### Mỗi cột hiện đang có kiểu dữ liệu gì? Có cột nào có kiểu dữ liệu chưa phù hợp để có thể xử lý tiếp không?

In [6]:
for col_name, col_vals in cols.items():
    print(f'{col_name} {type(col_vals[0])}')

State <class 'str'>
Started on <class 'str'>
Completed <class 'str'>
Time taken <class 'str'>
Grade/10.00 <class 'str'>
Q. 1 /0.53 <class 'str'>
Q. 2 /0.53 <class 'str'>
Q. 3 /0.53 <class 'str'>
Q. 4 /0.53 <class 'str'>
Q. 5 /0.53 <class 'str'>
Q. 6 /0.53 <class 'str'>
Q. 7 /0.53 <class 'str'>
Q. 8 /0.53 <class 'str'>
Q. 9 /0.53 <class 'str'>
Q. 10 /0.53 <class 'str'>
Q. 11 /0.53 <class 'str'>
Q. 12 /0.53 <class 'str'>
Q. 13 /0.53 <class 'str'>
Q. 14 /0.53 <class 'str'>
Q. 15 /0.53 <class 'str'>
Q. 16 /0.53 <class 'str'>
Q. 17 /0.53 <class 'str'>
Q. 18 /0.53 <class 'str'>
Q. 19 /0.53 <class 'str'>


Đúng như dự đoán, tất cả các cột đều đang có kiểu dữ liệu là str. Để có thể tiếp tục khám phá tiếp thì ta cần tiền xử lý để các cột có kiểu dữ liệu phù hợp.

- Cột "State": string là đúng rồi, không cần chuyển
- Cột "Started on" và "Completed on": chuyển sang kiểu dữ liệu datetime
- Cột "Time taken": chuyển sang một con số thể hiện cho số phút
- Các cột "Grade/16.00", "Q. 1 /1.00", ...: chuyển sang kiểu dữ liệu số

## Tiền xử lý

In [7]:
new_cols = {}
for col_name, col_vals in cols.items():
    if '/' in col_name:
        new_col_vals = []
        for col_val in col_vals:
            try:
                new_col_vals.append(float(col_val))
            except:
                new_col_vals.append(0.0)
        new_cols[col_name] = new_col_vals
    elif col_name in ['Started on', 'Completed']:
        new_col_vals = []
        for col_val in col_vals:
            new_col_val = dt.datetime.strptime(col_val, '%d/%m/%Y %H:%M')
            new_col_vals.append(new_col_val)
        new_cols[col_name] = new_col_vals
    elif col_name == 'Time taken':
        new_col_vals = []
        for col_val in col_vals:
            hour = 0; minute = 0; second = 0
            temp_list = col_val.split(' ')
            for temp_i in range(len(temp_list)):
                if 'hour' in temp_list[temp_i]:
                    hour = int(temp_list[temp_i - 1])
                elif 'min' in temp_list[temp_i]:
                    minute = int(temp_list[temp_i - 1])
                elif 'sec' in temp_list[temp_i]:
                    second = int(temp_list[temp_i - 1])
            new_col_val = hour * 60 + minute + second / 60
            new_col_vals.append(new_col_val)
        new_cols[col_name] = new_col_vals
    else:
        new_cols[col_name] = col_vals

In [8]:
cols = new_cols

In [9]:
for col_name, col_vals in cols.items():
    print(f'{col_name} {type(col_vals[0])}')

State <class 'str'>
Started on <class 'datetime.datetime'>
Completed <class 'datetime.datetime'>
Time taken <class 'float'>
Grade/10.00 <class 'float'>
Q. 1 /0.53 <class 'float'>
Q. 2 /0.53 <class 'float'>
Q. 3 /0.53 <class 'float'>
Q. 4 /0.53 <class 'float'>
Q. 5 /0.53 <class 'float'>
Q. 6 /0.53 <class 'float'>
Q. 7 /0.53 <class 'float'>
Q. 8 /0.53 <class 'float'>
Q. 9 /0.53 <class 'float'>
Q. 10 /0.53 <class 'float'>
Q. 11 /0.53 <class 'float'>
Q. 12 /0.53 <class 'float'>
Q. 13 /0.53 <class 'float'>
Q. 14 /0.53 <class 'float'>
Q. 15 /0.53 <class 'float'>
Q. 16 /0.53 <class 'float'>
Q. 17 /0.53 <class 'float'>
Q. 18 /0.53 <class 'float'>
Q. 19 /0.53 <class 'float'>


## Khám phá dữ liệu

### Với mỗi cột có kiểu dữ liệu dạng số (numerical), các giá trị được phân bố như thế nào?

- Số-lượng/tỉ-lệ các giá trị thiếu?
- Min? max? $\to$ Có gì bất thường không?

Ta thấy: cột "Started on" và "Completed" có kiểu dữ liệu datetime nhưng cũng có thể cho gia nhập hội numeric và tính min & max luôn.

Số-lượng/tỉ-lệ các giá trị thiếu? Không có giá trị thiếu.

Min? Max?

In [10]:
for col_name, col_vals in cols.items():
    if type(col_vals[0]) != str:
        print(f'{col_name}\t{min(col_vals)}\t{max(col_vals)}')

Started on	2024-10-16 17:14:00	2024-10-22 15:54:00
Completed	2024-10-16 17:37:00	2024-10-22 16:11:00
Time taken	5.45	215.0
Grade/10.00	2.24	10.0
Q. 1 /0.53	0.0	0.53
Q. 2 /0.53	0.0	0.53
Q. 3 /0.53	0.0	0.53
Q. 4 /0.53	0.0	0.53
Q. 5 /0.53	0.0	0.53
Q. 6 /0.53	0.0	0.53
Q. 7 /0.53	0.0	0.53
Q. 8 /0.53	0.13	0.53
Q. 9 /0.53	0.26	0.53
Q. 10 /0.53	0.26	0.53
Q. 11 /0.53	0.0	0.53
Q. 12 /0.53	0.26	0.53
Q. 13 /0.53	0.0	0.53
Q. 14 /0.53	0.0	0.53
Q. 15 /0.53	0.0	0.53
Q. 16 /0.53	0.0	0.53
Q. 17 /0.53	0.0	0.53
Q. 18 /0.53	0.0	0.53
Q. 19 /0.53	0.0	0.53


Có vẻ không có gì bất thường.

### Với mỗi cột có kiểu dữ liệu dạng phân loại (categorical), các giá trị được phân bố như thế nào?

- Số-lượng/tỉ-lệ các giá trị thiếu?
- Số lượng các giá trị khác nhau? Show một vài giá trị \
$\to$ Có gì bất thường không?

Hội này chỉ có một thành viên là cột "State"!

Số-lượng/tỉ-lệ các giá trị thiếu? Không có giá trị thiếu.

Số lượng các giá trị khác nhau? Show một vài giá trị

In [11]:
for col_name, col_vals in cols.items():
    if type(col_vals[0]) == str:
        col_vals_set = set(col_vals)
        print(f'{col_name}\t{len(col_vals_set)}\t{col_vals_set}')

State	1	{'Finished'}


Có vẻ không có gì bất thường.

## Đưa ra câu hỏi cần trả lời

Sau khi đã hiểu hơn về dữ liệu, ta sẽ quay lại bước đưa ra câu hỏi cần trả lời: tinh chỉnh câu hỏi, hoặc có thể đưa ra thêm các câu hỏi mới.

Câu hỏi “tình trạng hiểu về Python của các bạn trong lớp?” có thể được cụ thể hóa ra thành 2 câu hỏi:

1. Cột điểm tổng “Grade/16.00” có phân bố các giá trị như thế nào?
2. Câu quiz nào có nhiều bạn làm sai nhất, câu nào có nhiều bạn làm sai nhì, ...?

## Tiền xử lý + phân tích dữ liệu để trả lời cho câu hỏi 1

*Câu hỏi 1: Cột điểm tổng “Grade/10.00” có phân bố các giá trị như thế nào?*

Ở đây không cần tiền xử lý mà sẽ đi luôn vào phân tích dữ liệu.

Đầu tiên, ta sẽ xem thử median của cột "Grade/10.00".

In [12]:
grades = cols['Grade/10.00']
sorted_grades = sorted(grades)
p = 50
i = int(p/100 * (len(sorted_grades)-1))
sorted_grades[i]

8.68

Để cảm thụ rõ hơn về phân bố các giá trị của cột "Grade/10.00", ta sẽ chia miền giá trị của cột này thành các khoảng (bin) và xem thử mỗi khoảng có bao nhiêu sinh viên.

Giả sử ta sẽ chia đoạn [0, 10] thành 5 bin như sau: [0, 2], (2, 4], (4, 6], (6, 8], (8, 10]. Ta có thể lưu kết quả vào một dictionary, trong đó: key là một tuple cho biết 2 đầu của một bin (thay tuple bằng list thì có được không?), value là số lượng sinh viên rơi vào bin này.

***Sinh viên hoàn thành tiếp phần này như bài tập về nhà***

In [13]:
bin_ranges = [(0, 2), (2, 4), (4, 6), (6, 8), (8, 10)]
bin_counts = {bin_range: 0 for bin_range in bin_ranges}

grades = cols['Grade/10.00']
for grade in grades:
    for bin_range in bin_ranges:
        low, high = bin_range
        if (low == 0 and grade == low) or low < grade <= high:
            bin_counts[bin_range] += 1
            break

bin_counts

{(0, 2): 0, (2, 4): 1, (4, 6): 9, (6, 8): 9, (8, 10): 40}

**Trả lời câu hỏi 1.1:** *Cột điểm tổng “Grade/10.00” có phân bố các giá trị như thế nào?*

Kết quả thống kê cho thấy số lượng sinh viên trong mỗi bin như sau:
- Từ 0 đến 2: 0 sinh viên
- Từ 2 đến 4: 1 sinh viên
- Từ 4 đến 6: 9 sinh viên
- Từ 6 đến 8: 9 sinh viên
- Từ 8 đến 10: 40 sinh viên

Kết quả trên cho thấy hầu hết sinh viên có điểm từ 8 đến 10, nghĩa là phần lớn điểm tập trung ở mức cao, qua đó có thể đánh giá tích cực về kết quả của bài kiểm tra.

**Trả lời câu hỏi 1.2:** *Thay tuple bằng list thì có được không?*

Ta không thể sử dụng `list` thay `tuple` để làm `key` trong `dictionary` với những lý do sau đây: 
- `key` phải là kiểu dữ liệu bất biến `immutable` và có thể hash `hashable`. Do đó ta cần chọn một kiểu dữ liệu `immutable` tương ứng để làm `key`.
- Theo tính chất của `key`, ta không thể chọn `list` vì list là kiểu dữ liệu có thể thay đổi `mutable` và không thể hash `unhashable`.
- `tuple` là một kiểu dữ bất biến `immutable` và có thể hash `hashable` đáp ứng yêu cầu của `key`.

Vì vậy, ta không thể thay `tuple` bằng `list`. Sử dụng `tuple` làm `key` trong `dictionary` sẽ đảm bảo rằng giá trị của `key` là bất biến, một `key` sẽ được hash thành một mã duy nhất không thể thay đổi. Điều này giúp dữ liệu đáng tin cậy và an toàn hơn.

Tuy không thể sử dụng trực tiếp `list` để làm `key`, ta vẫn có thể sử dụng gián tiếp bằng cách chuyển `list` sang `string`, `frozenset` hoặc một kiểu dữ liệu `immutable` khác. Nhưng cũng cần lưu ý rằng việc này có thể làm phức tạp đoạn mã và khó quản lý hơn.

## Tiền xử lý + phân tích dữ liệu để trả lời cho câu hỏi 2

*Câu hỏi 2: Câu quiz nào có nhiều bạn làm sai nhất, câu nào có nhiều bạn làm sai nhì, ...?*

Ta có thể lưu kết quả vào một dictionary, trong đó: key là tên của cột ứng với câu quiz, value là điểm trung bình của cột. Dictionary này được sort tăng dần theo value.

***Sinh viên hoàn thành tiếp phần này như bài tập về nhà***

In [14]:
fail_counts = {}
average_grades = {}

for question, grades in cols.items():
    if 'Q.' in question:
        fail_count = 0
        total_score = 0
        for grade in grades:
            total_score += grade
            if grade < 0.53:
                fail_count += 1
        fail_counts[question] = fail_count
        average_grades[question] = total_score / len(grades)

In [15]:
# Sắp xếp theo số câu sai giảm dần để tương ứng với số điểm trung bình tăng dần
sorted_fail_counts = sorted(fail_counts.items(), key=lambda item: item[1], reverse=True)

print('Câu hỏi \t Số câu sai \n' + '-'*27)
for question, count in sorted_fail_counts:
    print(f'{question} \t {count}')

Câu hỏi 	 Số câu sai 
---------------------------
Q. 1 /0.53 	 22
Q. 5 /0.53 	 21
Q. 8 /0.53 	 20
Q. 13 /0.53 	 19
Q. 14 /0.53 	 19
Q. 18 /0.53 	 18
Q. 6 /0.53 	 17
Q. 10 /0.53 	 16
Q. 11 /0.53 	 16
Q. 19 /0.53 	 12
Q. 4 /0.53 	 11
Q. 17 /0.53 	 10
Q. 3 /0.53 	 9
Q. 7 /0.53 	 9
Q. 15 /0.53 	 9
Q. 9 /0.53 	 7
Q. 12 /0.53 	 6
Q. 16 /0.53 	 4
Q. 2 /0.53 	 2


In [16]:
# điểm trung bình được sắp xếp tăng dần
sorted_average_grades = sorted(average_grades.items(), key=lambda item: item[1])

print('Câu hỏi \t Điểm trung bình \n' + '-'*36)
for question, avg in sorted_average_grades:
    print(f'{question} \t {avg}')

Câu hỏi 	 Điểm trung bình 
------------------------------------
Q. 5 /0.53 	 0.34135593220338983
Q. 13 /0.53 	 0.35932203389830514
Q. 14 /0.53 	 0.35932203389830514
Q. 18 /0.53 	 0.3683050847457628
Q. 6 /0.53 	 0.37728813559322044
Q. 1 /0.53 	 0.4116949152542375
Q. 19 /0.53 	 0.42220338983050865
Q. 4 /0.53 	 0.4311864406779663
Q. 11 /0.53 	 0.43915254237288165
Q. 17 /0.53 	 0.44016949152542395
Q. 3 /0.53 	 0.4491525423728816
Q. 7 /0.53 	 0.4491525423728816
Q. 15 /0.53 	 0.4491525423728816
Q. 8 /0.53 	 0.45610169491525454
Q. 10 /0.53 	 0.4832203389830512
Q. 16 /0.53 	 0.4940677966101698
Q. 9 /0.53 	 0.4979661016949156
Q. 12 /0.53 	 0.5025423728813563
Q. 2 /0.53 	 0.5120338983050852


**Trả lời câu hỏi 2:** *Câu quiz nào có nhiều bạn làm sai nhất, câu nào có nhiều bạn làm sai nhì, ...?*

1. Số lượng sinh viên làm sai mỗi câu (giảm dần theo số lượng làm sai): Q.1 có số lượng sinh viên làm sai nhiều nhất (22 sinh viên), tiếp theo là Q.5 (21 sinh viên) và dần xuống.

2. Điểm trung bình mỗi câu (tăng dần theo điểm trung bình): Q.5 có điểm trung bình thấp nhất (0.34), Q.13 và Q.14 có điểm trung bình thấp nhì (0.359) và dần tăng.

**Lưu ý:** Do trong bài kiểm tra có câu hỏi multiple-choice, những câu không đạt điểm tối đa sẽ được tính là sai. Điều này giải thích tại sao số lượng sinh viên làm sai có thể không hoàn toàn tương ứng với điểm trung bình.